File: sorting/c/quick_sort.c
#include "swap.h"

void quick_sort(int* array, const int n);
void q_sort(int* array, int left, int right);
int partition(int* array, int left, int right);
int median3(int *array, int left, int right);
void _insertion_sort(int* array, int left, int right);

void _insertion_sort(int* array, int left, int right)
{
	int i, j, x;
	for(i = left + 1; i < right + 1; i++)
	{
		x = array[i];
		for(j = i - 1; j >= left; j--)
			if(array[j] > x)
				array[j + 1] = array[j];
			else
				break;
		array[j + 1] = x;
	}
}

void quick_sort(int* array, const int n)
{
	q_sort(array, 0, n - 1);
}

void q_sort(int* array, int left, int right)
{
	if(right - left > 10)
	{
		int center = partition(array, left, right);
		q_sort(array, left, center - 1);
		q_sort(array, center + 1, right);
	}
	else
		_insertion_sort(array, left, right);
}

int median3(int* array, int left, int right)
{
	int center = (left + right) / 2;
	int tmp[] = { array[left], array[center], array[right] };
	_insertion_sort(tmp, 0, 2);
	array[left] = tmp[0];
	array[center] = tmp[2];
	array[right] = tmp[1];
	return array[right];
}

int partition(int* array, int left, int right)
{
	int pivot = median3(array, left, right);
	int i = left + 1, j = right - 1;
	while(1)
	{
		while(array[i] < pivot) i++;
		while(pivot < array[j]) j--;
		if(i >= j)
		{
			swap(&array[i], &array[right]);
			return i;
		}
		swap(&array[ i++ ], &array[ j-- ]);
	}
}

File: sorting/go/quicksort.go
package main

// TESTED on LeetCode

// QuickSort sorts the array
func QuickSort(nums []int) {
	qSort(nums, 0, len(nums)-1)
}

func qSort(nums []int, left, right int) {
	if right-left > 10 {
		center := partition(nums, left, right)
		qSort(nums, left, center-1)
		qSort(nums, center+1, right)
	} else {
		insertionSort(nums, left, right)
	}
}

func insertionSort(nums []int, left, right int) {
	var j, x int
	for i := left + 1; i < right+1; i++ {
		x = nums[i]
		for j = i - 1; j >= left && nums[j] > x; j-- {
			nums[j+1] = nums[j]
		}
		nums[j+1] = x
	}
}

func median3(nums []int, left, right int) int {
	center := left + (right-left)/2
	tmp := []int{nums[left], nums[center], nums[right]}
	insertionSort(tmp, 0, 2)
	nums[left], nums[center], nums[right] = tmp[0], tmp[2], tmp[1]
	return nums[right]
}

func partition(nums []int, left, right int) int {
	pivot := median3(nums, left, right)
	i, j := left+1, right-1
	for {
		for nums[i] < pivot {
			i++
		}
		for pivot < nums[j] {
			j--
		}
		if i >= j {
			nums[i], nums[right] = nums[right], nums[i]
			return i
		}
		nums[i], nums[j] = nums[j], nums[i]
		i++
		j--
	}
}
